smtest/event_manager.go

```diff
@@ -151,32 +148,32 @@ func (ei *EventIterator) SkipTill(pred EventP) {
 
 // TakeTill takes all the events that have been collected since the current
 // index until the given predicate returns true
-func (ei *EventIterator) TakeTill(pred EventP) []cap.Event {
-	zero := make([]cap.Event, 0, 100)
-	iresult := ei.foldl(zero, func(iacc interface{}, ev cap.Event) (bool, interface{}) {
-		acc, _ := iacc.([]cap.Event)
+func (ei *EventIterator[A]) TakeTill(pred EventP[A]) []A {
+	zero := make([]A, 0, 100)
+	iresult := ei.foldl(zero, func(iacc interface{}, ev A) (bool, interface{}) {
+		acc, _ := iacc.([]A)
 		if pred.Call(ev) {
 			return false, acc
 		}
 		acc = append(acc, ev)
 		return true, acc
 	})
-	result, _ := iresult.([]cap.Event)
+	result, _ := iresult.([]A)
```

Comment:
Is there a way to use generics and avoid the cast?

Reply:
We would need to change `foldl` from a method to a function given that methods cannot have type parameters :sad-gonzo:, let me try to do that.